### ISSUE TITLE
v7.0.3 throws TypeScript error when adding ajv-formats (works in v7.0.2)

### ISSUE THREAD

**Issue by 1valdis** [ajv-validator/ajv#1378]
**What version of Ajv are you using? Does the issue happen if you use the latest version?**
I use `ajv@7.0.2` and it works with `ajv-formats@1.5.1`, but when I upgrade ajv to `7.0.3`, it breaks.

**Your typescript code**

```typescript
import Ajv from 'ajv';
import addFormats from 'ajv-formats';

const validator = new Ajv({
  allErrors: true,
});

addFormats(validator); // error is here

export default validator;
```

**Typescript compiler error messages**

```
Argument of type 'import("<project directory>/node_modules/ajv/dist/ajv").default' is not assignable to parameter of type 'import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/core").default'.
  Types of property 'opts' are incompatible.
    Type 'import("<project directory>/node_modules/ajv/dist/core").InstanceOptions' is not assignable to type 'import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/core").InstanceOptions'.
      Type 'InstanceOptions' is not assignable to type 'CurrentOptions'.
        Types of property 'keywords' are incompatible.
          Type 'import("<project directory>/node_modules/ajv/dist/types/index").Vocabulary | undefined' is not assignable to type 'import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/types/index").Vocabulary | undefined'.
            Type 'import("<project directory>/node_modules/ajv/dist/types/index").Vocabulary' is not assignable to type 'import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/types/index").Vocabulary'.
              Type 'string | import("<project directory>/node_modules/ajv/dist/types/index").CodeKeywordDefinition | import("<project directory>/node_modules/ajv/dist/types/index").FuncKeywordDefinition | import("<project directory>/node_modules/ajv/dist/types/index").MacroKeywordDefinition' is not assignable to type 'string | import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/types/index").CodeKeywordDefinition | import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/types/index").FuncKeywordDefinition | import("<project directory>/node_modules/aj...'.
                Type 'CodeKeywordDefinition' is not assignable to type 'string | CodeKeywordDefinition | FuncKeywordDefinition | MacroKeywordDefinition'.
                  Type 'import("<project directory>/node_modules/ajv/dist/types/index").CodeKeywordDefinition' is not assignable to type 'import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/types/index").CodeKeywordDefinition'.
                    Types of property 'code' are incompatible.
                      Type '(cxt: import("<project directory>/node_modules/ajv/dist/compile/context").default, ruleType?: string | undefined) => void' is not assignable to type '(cxt: import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/compile/context").default, ruleType?: string | undefined) => void'.
                        Types of parameters 'cxt' and 'cxt' are incompatible.
                          Type 'import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/compile/context").default' is not assignable to type 'import("<project directory>/node_modules/ajv/dist/compile/context").default'.
                            The types of 'gen._scope' are incompatible between these types.
                              Type 'import("<project directory>/node_modules/ajv-formats/node_modules/ajv/dist/compile/codegen/scope").Scope' is not assignable to type 'import("<project directory>/node_modules/ajv/dist/compile/codegen/scope").Scope'.
                                Property '_names' is protected but type 'Scope' is not a class derived from 'Scope'.ts(2345)
```

**Describe the change that should be made to address the issue?**
Locally, I reverted back to `ajv@7.0.2`. Adding `// @ts-ignore` to the line before the error while using `7.0.3` works too. I don't know how to fix this otherwise.

**Are you going to resolve the issue?**
No, I don't know how and don't have time to dig into.

**Comment by epoberezkin** [comment]
Can you please try deleting node_modules and reinstalling everything?

**Comment by 1valdis** [comment]
@epoberezkin yep I did it and got the same result. I deleted node_modules, set ajv to `=7.0.3` in package.json, installed and got the error. Then I deleted node_modules again, set ajv to `=7.0.2`, installed and got no error. ajv-formats was set to `=1.5.1` the entire time.

Some other info which could be useful: I work on `typescript@4.1.3`, and have the following tsconfig.json:
```json
{
  "compilerOptions": {
    "module": "commonjs",
    "declaration": true,
    "removeComments": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "allowSyntheticDefaultImports": true,
    "target": "es2020",
    "sourceMap": true,
    "outDir": "./dist",
    "baseUrl": "./",
    "incremental": true,
    "strict": true,
    "noUncheckedIndexedAccess": true,
    "esModuleInterop": true
  },
  "exclude": ["node_modules", "dist"]
}
```

**Comment by epoberezkin** [comment]
I can't reproduce I am afraid.

I just created a clean installation with your code, your tsconfig and this is package.json:

```json
{
  "name": "ajv-test",
  "version": "0.0.1",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "ajv": "^7.0.3",
    "ajv-formats": "^1.5.1"
  },
  "devDependencies": {
    "typescript": "^4.1.3"
  }
}
```

it compiles and executes... Can you make a sample repo that reproduces the problem?

```
node -v 
v14.14.0
npm -v
6.14.10
```

Maybe you have older typesript version?

**Comment by 1valdis** [comment]
@epoberezkin I couldn't reproduce while simply making empty project and adding ajv there, just like you, too, so I instead copied my project with error and stripped nearly everything away, and the error was still there. Then I deleted `package-lock.json` along with `node_modules`, and after doing `npm i` there was no error. So I compared the old `package-lock.json` which caused the error to be there and the one which was newly generated. I found that there was a difference in `ajv-formats` entry.
The one which produced no error:
```json
"ajv-formats": {
      "version": "1.5.1",
      "resolved": "https://registry.npmjs.org/ajv-formats/-/ajv-formats-1.5.1.tgz",
      "integrity": "sha512-s1RBVF4HZd2UjGkb6t6uWoXjf6o7j7dXPQIL7vprcIT/67bTD6+5ocsU0UKShS2qWxueGDWuGfKHfOxHWrlTQg==",
      "requires": {
        "ajv": "^7.0.0"
      }
    },
```
The one which did produce an error:
```json
"ajv-formats": {
      "version": "1.5.1",
      "resolved": "https://registry.npmjs.org/ajv-formats/-/ajv-formats-1.5.1.tgz",
      "integrity": "sha512-s1RBVF4HZd2UjGkb6t6uWoXjf6o7j7dXPQIL7vprcIT/67bTD6+5ocsU0UKShS2qWxueGDWuGfKHfOxHWrlTQg==",
      "requires": {
        "ajv": "^7.0.0"
      },
      "dependencies": {
        "ajv": {
          "version": "7.0.2",
          "resolved": "https://registry.npmjs.org/ajv/-/ajv-7.0.2.tgz",
          "integrity": "sha512-qIaxO9RXjaXyO21tJr0EvwPcZa49m64GcXCU8ZrLjCAlFyMuOcPpI560So4A11M1WsKslJYIXn6jSyG5P0xIeg==",
          "requires": {
            "fast-deep-equal": "^3.1.1",
            "json-schema-traverse": "^1.0.0",
            "require-from-string": "^2.0.2",
            "uri-js": "^4.2.2"
          }
        },
        "json-schema-traverse": {
          "version": "1.0.0",
          "resolved": "https://registry.npmjs.org/json-schema-traverse/-/json-schema-traverse-1.0.0.tgz",
          "integrity": "sha512-NM8/P9n3XjXhIZn1lLhkFaACTOURQXjWhV4BA/RnOv8xvgqtqpAX9IO4mRQxSx1Rlo4tqzeqb0sOlruaOy3dug=="
        }
      }
    },
```
So I then did just the same in my main project: deleted not only `node_modules` but `package-lock.json` too, reinstalled, and it worked. Previously I kept `package-lock.json` in place and it didn't. I don't have a clue why `npm` decided to setup ajv-formats' dependencies this weird way.

Thank you for attention, I'm sorry I took your time, keep it up with the best json schema validation package out there :)

**Comment by epoberezkin** [comment]
No problem, and thanks for writing it up - it might help other people.

Yes, package-lock.json can indeed lock sub-dependencies to the old incompatible versions - I will remember it :)

**Comment by mdurling** [comment]
I am experiencing this exact same problem (using `yarn`). I can sometimes get it to a point where `tsc` will compile, but `webpack` still reports the error. Tried deleting `yarn.lock`, etc. to no avail. The only option I have is to set `ts-loader` to `transpileOnly: true`, but this all makes me nervous.

**Comment by epoberezkin** [comment]
It looks like every time the new version is released it happens - so yarn will continue to be affected until all the dependencies are aligned on the same major version (that can take quite long)...

My recommendation is to use npm - it seems at least as fast now (if not faster) and definitely more stable at managing complex dependency trees.

This issue should be reported to yarn - it should be able to handle multiple versions of the same package used by different dependencies of your package.

**Comment by huantaoliu** [comment]
Hi @epoberezkin can you explain a little bit more on why everytime a new version of AJV is realesed, the problem will happen on Yarn? 2 days ago, Ajv released 8.14.0. I got the above error again. I guess the problem is that when sub lib require ajv by a range of version like `^8.0.0`, yarn will resolve them all to the latest `8.14.0` and install it. And in my packge.json I specified to use `8.13.0`, So I guess the incompatiable error is between 8.13.0 and the acturally resolved new one 8.14.0? But when i check the new release, 8.14.0 doesn't have big code difference between 8.13.0, why comes the lib incompatible error ?

**Comment by jgburet** [comment]
Same issue here using pnpm
